crates/sui/src/client_ptb/builder.rs

```diff
@@ -0,0 +1,993 @@
+            }
+            PTBArg::Identifier(ref i) => {
+                // We do a one-hop resolution here to see if we can resolve the identifier to an
+                // externally-bound address (i.e., one coming in through the initial environment).
+                // This will also handle direct aliasing of addresses throughout the ptb.
+                // Note that we don't do this recursively so no need to worry about loops/cycles.
+                if let Some(addr) = self.addresses.get(i) {
+                    self.addresses.insert(ident, *addr);
+                }
+            }
+            // If we encounter a dotted string e.g., "foo.0" or "sui.io" or something like that
+            // this see if we can find an address for it in the environment and bind to it.
+            PTBArg::VariableAccess(ref head, ref fields) => {
+                let key = format!(
+                    "{}.{}",
+                    head.value,
+                    fields
+                        .iter()
+                        .map(|f| f.value.clone())
+                        .collect::<Vec<_>>()
+                        .join(".")
+                );
+                if let Some(addr) = self.addresses.get(&key) {
+                    self.addresses.insert(ident, *addr);
+                }
+            }
+            _ => (),
+        }
+    }
+
+    async fn get_protocol_config(&self, loc: Span) -> PTBResult<ProtocolConfig> {
+        let config = self
+            .reader
+            .get_protocol_config(None)
+            .await
+            .map_err(|e| err!(loc, "{e}"))?;
+        let chain_id = self
+            .reader
+            .get_chain_identifier()
+            .await
+            .map_err(|e| err!(loc, "{e}"))?;
+        Ok(if chain_id == get_mainnet_chain_identifier().to_string() {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Mainnet)
+        } else if chain_id == get_testnet_chain_identifier().to_string() {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Testnet)
+        } else {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Unknown)
+        })
+    }
+
+    /// Resolve an object ID to a Move package.
+    async fn resolve_to_package(
+        &mut self,
+        package_id: ObjectID,
+        loc: Span,
+    ) -> PTBResult<MovePackage> {
+        let object = self
+            .reader
+            .get_object_with_options(package_id, SuiObjectDataOptions::bcs_lossless())
+            .await
+            .map_err(|e| err!(loc, "{e}"))?
+            .into_object()
+            .map_err(|e| err!(loc, "{e}"))?;
+        let Some(SuiRawData::Package(package)) = object.bcs else {
+            error!(
+                loc,
+                "BCS field in object '{}' is missing or not a package.", package_id
+            );
+        };
+        let protocol_config = self.get_protocol_config(loc).await?;
```

Comment:
This was confusing, the loc is needed for errors and not for actually anything related to the protocol. Need to get used to these spans everywhere.